### PR TITLE
Add rotation limit to unit weapons

### PR DESCRIPTION
**Pull request by z0mbiesrock**
This adds a variable to weapons that limits their relative angle from their initial rotation. This can prevent a mech's "arms" from rotating at awkward angles.

### CODE REVIEW

**Review by z0mbiesrock — COMMENTED** [review]
holy f:ohno: I finally solved it

image

File: core/src/mindustry/entities/comp/WeaponsComp.java [hunk 1]
```diff
@@ -148,8 +148,13 @@ public void update(){
             if(weapon.rotate && (mount.rotate || mount.shoot) && can){
                 float axisX = this.x + Angles.trnsx(this.rotation - 90,  weapon.x, weapon.y),
                     axisY = this.y + Angles.trnsy(this.rotation - 90,  weapon.x, weapon.y);
-
                 mount.targetRotation = Angles.angle(axisX, axisY, mount.aimX, mount.aimY) - this.rotation;
+                if(weapon.rotateLimit > 0){
+                    mount.rotation = Angles.moveToward(mount.rotation, Mathf.clamp(mount.targetRotation, -(weapon.rotateLimit), weapon.rotateLimit), weapon.rotateSpeed * Time.delta);
+                }
+                else{
```
Comment (joshuaptfan) [inline core/src/mindustry/entities/comp/WeaponsComp.java:155]: Style. This should be 1 line.

**Comment by DeltaNedas** [comment]
may want to squash those 19 commits chief

**Comment by z0mbiesrock** [comment]
> may want to squash those 19 commits chief
That's what Anuke does when he merges pull requests.

File: core/src/mindustry/entities/comp/WeaponsComp.java [hunk 2]
```diff
@@ -148,8 +148,13 @@ public void update(){
...
+                    mount.rotation = Angles.moveToward(mount.rotation, mount.targetRotation, weapon.rotateSpeed * Time.delta);
+                }
```
Comment (MEEPofFaith) [inline core/src/mindustry/entities/comp/WeaponsComp.java:157]: > Style. This should be 1 line.

```suggestion
                if(weapon.rotateLimit > 0) mount.rotation = Angles.moveToward(mount.rotation, Mathf.clamp(mount.targetRotation, -(weapon.rotateLimit), weapon.rotateLimit), weapon.rotateSpeed * Time.delta);
                else mount.rotation = Angles.moveToward(mount.rotation, mount.targetRotation, weapon.rotateSpeed * Time.delta);
```
(Does this work?)

Comment (joshuaptfan) [reply core/src/mindustry/entities/comp/WeaponsComp.java:157]: No, I only mean
```
}
else{
```
should be one line, like
```
}else{
```

**Comment by DeltaNedas** [comment]
bruh you make yourself look like an idiot having 20 commits of just testing things

File: core/src/mindustry/entities/comp/WeaponsComp.java (hunk 2 above)
Comment (MEEPofFaith) [reply core/src/mindustry/entities/comp/WeaponsComp.java:157]: ah ok

**Comment by MEEPofFaith** [comment]
`git merge --squash`

**Comment by z0mbiesrock** [comment]
> `git merge --squash`

which means I have to install Git, clone the branch, and use the terminal to merge using Git?

No seriously @Anuken can just squash the commits as he merges this pr though.

**Comment by MEEPofFaith** [comment]
no need to clone but ok

**Comment by ThatOneBepis** [comment]
To be honest, installing git isn't that much of an issue considering help it can provide.

File: core/src/mindustry/type/Weapon.java [hunk 3]
```diff
@@ -32,6 +32,8 @@
     public boolean continuous;
     /** rotation speed of weapon when rotation is enabled, in degrees/t*/
     public float rotateSpeed = 20f;
+    /** rotation speed of weapon when rotation is enabled, in degrees/t*/
```
Comment (Quezler) [inline core/src/mindustry/type/Weapon.java:35]:
```suggestion
    /** rotation speed of weapon when rotation is enabled, in degrees/t */
```

**Comment by BlueWolf3682** [comment]
ngl those reigns look mighty sus

**Comment by XEN0PHIL** [comment]
nice, adding yourself to the contributor list before contributing

**Comment by DeltaNedas** [comment]
bruh

**Comment by summetdev** [comment]
> nice, adding yourself to the contributor list before contributing

> If you would like your name to appear in the game's credits, add it to the list of contributors as part of your PR.


In CONTRIBUTING.md

**Comment by Anuken** [comment]
Tested this out, and it just looks super janky. Weapons also tend to twitch around and assume impossible angles.

**Comment by BlueWolf3682** [comment]
yay no sus reigns

**Pull request closed without merging**